File: Classes/Controls/DocumentView/Files/ImageView/FileImageView.m
//
//  FileImageView.m
//  Dokie
//
//  Created by Ashamaz Shidov on 26.01.2020.
//  Copyright © 2020 Ashamaz Shidov. All rights reserved.
//

#import "FileImageView.h"
#import "CoreDataManager.h"

#import <ReactiveObjC/ReactiveObjC.h>

@interface FileImageView()

@property (nonatomic, strong) RACSubject *stopSubject;
@property (nonatomic, strong, readonly) CoreDataManager *coreDataManager;

@end

@implementation FileImageView

- (void)updateWithContent:(CDContent *)content {
    self.image = nil;
    [self.stopSubject sendNext:nil];
    @weakify(self);
    [[[[self.coreDataManager imageFromContent:content]
       takeUntil:self.stopSubject]
      deliverOnMainThread]
     subscribeNext:^(UIImage *image) {
        @strongify(self);
        self.image = image;
    }];
}

- (RACSubject *)stopSubject {
    if (!_stopSubject) {
        _stopSubject = [RACSubject new];
    }
    return _stopSubject;
}

- (CoreDataManager *)coreDataManager {
    return [CoreDataManager shared];
}

@end

File: Classes/Views/DocumentView/Files/ImageView/FileImageView.swift
//
//  FileImageView.swift
//  Dokie
//
//  Created by Ashamaz Shidov on 24.05.2020.
//  Copyright © 2020 Ashamaz Shidov. All rights reserved.
//

import UIKit
import ReactiveCocoa
import ReactiveSwift

class FileImageView: UIImageView {
    
    private let coreDataManager = CoreDataManager.shared
    private var disposable: Disposable?
    
    func update(content: CDContent) {
        image = nil
        disposable?.dispose()
        disposable = coreDataManager
            .imageFrom(content: content)
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithResult { [unowned self] result in
                switch result {
                case .success(let image): self.image = image
                case .failure: self.image = nil
                }
        }
    }
    
}
